src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs

```diff
@@ -315,6 +320,227 @@ public static bool IsUnnecessaryAsCast(BinaryExpressionSyntax cast, SemanticMode
             return false;
         }
 
+        private static bool CastRemovalWouldCauseSignExtensionWarning(ExpressionSyntax expression, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // Logic copied from DiagnosticsPass_Warnings.CheckForBitwiseOrSignExtend.  Including comments.
```

Comment:
Would it be possible to move this logic to a shared helper class in a separate file in compiler layer, so it can be linked into this project instead of cloned? I am concerned this would diverge pretty soon and introduce bugs.

Reply:
I understand this has lot of customer logic specific to the analyzer, but it would be good if at least the core logic that can possibly be shared is shared. Please let me know if you have already done so for everything possible, in which case it is fine.

Reply:
1. Unfortunately, the compiler logic is all built on compiler internals (i.e. bound nodes).
2. i did have to tweak some things.  for example, our function says: would this be an issue if the cast was *removed*.  the compiler version obviously is written analyzing the actual expression witht the cast.

> I am concerned this would diverge pretty soon and introduce bugs.

So i looked, and this hasn't changed in the compiler layer since we went open source.  it's effectively a very very old holdover from the native compiler.  So i'm not really concerned about that.

> Please let me know if you have already done so for everything possible, in which case it is fine.

I tried to share as much as i could.  all the helper extensions are shared.  but the core logic that does this specific check is still a copy :-/